zkevm-circuits/src/tx_circuit/sign_verify.rs

```diff
@@ -0,0 +1,978 @@
+// Naming notes:
+// - *_be: Big-Endian bytes
+// - *_le: Little-Endian bytes
+
+use crate::{
+    evm_circuit::util::{not, RandomLinearCombination, Word},
+    util::Expr,
+};
+use ecc::{EccConfig, GeneralEccChip};
+use ecdsa::ecdsa::{AssignedEcdsaSig, AssignedPublicKey, EcdsaChip};
+use gadgets::is_zero::{IsZeroChip, IsZeroConfig, IsZeroInstruction};
+use group::{ff::Field, prime::PrimeCurveAffine, Curve};
+use halo2_proofs::{
+    arithmetic::{BaseExt, Coordinates, CurveAffine, FieldExt},
+    circuit::{AssignedCell, Layouter, Region},
+    plonk::{Advice, Column, ConstraintSystem, Error, Expression, Selector},
+    poly::Rotation,
+};
+use integer::{
+    AssignedInteger, IntegerChip, IntegerConfig, IntegerInstructions, WrongExt,
+    NUMBER_OF_LOOKUP_LIMBS,
+};
+use itertools::Itertools;
+use keccak256::plain::Keccak;
+use lazy_static::lazy_static;
+use log::error;
+use maingate::{
+    Assigned, AssignedValue, MainGate, MainGateConfig, MainGateInstructions, RangeChip,
+    RangeConfig, RangeInstructions, RegionCtx, UnassignedValue,
+};
+use secp256k1::Secp256k1Affine;
+use std::{convert::TryInto, io::Cursor, marker::PhantomData};
+
+/// Power of randomness vector size required for the SignVerifyChip
+pub const POW_RAND_SIZE: usize = 63;
+
+/// Number of rows required for a verification of the SignVerifyChip in the
+/// "signature address verify" region.
+pub const VERIF_HEIGHT: usize = 1;
+
+/// Auxiliary Gadget to verify a that a message hash is signed by the public
+/// key corresponding to an Ethereum Address.
+#[derive(Default, Debug)]
+pub(crate) struct SignVerifyChip<F: FieldExt, const MAX_VERIF: usize> {
+    pub aux_generator: Secp256k1Affine,
+    pub window_size: usize,
+    pub _marker: PhantomData<F>,
+}
+
+const KECCAK_IS_ENABLED: usize = 0;
+const KECCAK_INPUT_RLC: usize = 1;
+const KECCAK_INPUT_LEN: usize = 2;
+const KECCAK_OUTPUT_RLC: usize = 3;
+
+const NUMBER_OF_LIMBS: usize = 4;
+const BIT_LEN_LIMB: usize = 72;
+
+/// Return an expression that builds an integer element in the field from the
+/// `bytes` in little endian.
+fn int_from_bytes_le<'a, F: FieldExt>(
+    bytes: impl IntoIterator<Item = &'a Expression<F>>,
+) -> Expression<F> {
+    // sum_{i = 0}^{N} bytes[i] * 256^i
+    let mut res = 0u8.expr();
+    for (i, byte) in bytes.into_iter().enumerate() {
+        res = res + byte.clone() * Expression::Constant(F::from(256).pow(&[i as u64, 0, 0, 0]))
```

Comment:
What about precompute them in a `lazy_static`?

Reply:
Thanks for the suggestion.  I've taken a look and the function `int_from_bytes_le` is only called once (during the chip config), so I would say there's no need to put it int a `lazy_static` because the computation is not going to be reused.

On the other hand there is this: https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/e10a0eb379ea9d098cc0e521c0100842c42183cc/zkevm-circuits/src/tx_circuit/sign_verify.rs#L396
Which is called for every signature.  These constants are used to convert an emulated integer inside the maingate (the halo2wrong gate) into bytes.  Nevertheless I plan to remove this in favor of a halo2wrong chip that performs the conversion to bytes more efficiently; and for that reason I think it's not worth moving it to `lazy_static` now.

Reply:
+1